openmeter/ent/schema/subscription.go

```diff
@@ -0,0 +1,135 @@
+package schema
+
+import (
+	"entgo.io/ent"
+	"entgo.io/ent/dialect"
+	"entgo.io/ent/schema/edge"
+	"entgo.io/ent/schema/field"
+	"entgo.io/ent/schema/index"
+
+	"github.com/openmeterio/openmeter/openmeter/productcatalog/plan"
+	"github.com/openmeterio/openmeter/pkg/currencyx"
+	"github.com/openmeterio/openmeter/pkg/framework/entutils"
+)
+
+type Subscription struct {
+	ent.Schema
+}
+
+func (Subscription) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
+		entutils.CadencedMixin{},
+	}
+}
+
+func (Subscription) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("plan_key").NotEmpty().Immutable(),
+		field.Int("plan_version").Min(1).Immutable(),
+		field.String("customer_id").NotEmpty().Immutable(),
+		field.String("currency").GoType(currencyx.Code("")).MinLen(3).MaxLen(3).NotEmpty().Immutable(),
+	}
+}
+
+func (Subscription) Indexes() []ent.Index {
+	return []ent.Index{
+		index.Fields("namespace", "id"),
+		index.Fields("namespace", "customer_id"),
+	}
+}
+
+func (Subscription) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.To("prices", Price.Type),
+		edge.To("entitlements", SubscriptionEntitlement.Type),
+		edge.From("customer", Customer.Type).Field("customer_id").Ref("subscription").Immutable().Unique().Required(),
+		edge.To("phases", SubscriptionPhase.Type),
+	}
+}
+
+type SubscriptionPhase struct {
+	ent.Schema
+}
+
+func (SubscriptionPhase) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
+	}
+}
+
+func (SubscriptionPhase) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("subscription_id").NotEmpty().Immutable(),
+		field.String("key").NotEmpty().Immutable(),
+		field.String("name").NotEmpty(),
+		field.String("description").Optional().Nillable(),
+		field.Time("active_from").Immutable(),
+	}
+}
+
+func (SubscriptionPhase) Indexes() []ent.Index {
+	return []ent.Index{
+		index.Fields("namespace", "id"),
+		index.Fields("namespace", "subscription_id"),
+		index.Fields("namespace", "subscription_id", "key"),
+	}
+}
+
+func (SubscriptionPhase) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.From("subscription", Subscription.Type).Field("subscription_id").Ref("phases").Unique().Immutable().Required(),
+		edge.To("items", SubscriptionItem.Type),
+	}
+}
+
+type SubscriptionItem struct {
+	ent.Schema
+}
+
+func (SubscriptionItem) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
```

Comment:
why do we need annotations instead of just adding a  readonly or editable flag like: `field.Bool("readonly").NotEmpty().Immutable().Default(false)` ?